### ISSUE TITLE
[Proposal] Tracing API and Interface for application level tracing components

### ISSUE THREAD

**Issue by yaron2** [dapr/dapr#100]
Goal: allow users to leverage distributed tracing from their code, regardless of whether the request came in through Dapr.

Here's a very initial rough draft to iterate on:

```
type Trace interface {
	TraceOperation(ctx context.Context, req TraceOperationRequest) error
}

type TraceOperationRequest struct {
	Operation string
	Data      interface{}
}
```

@Haishi2016

**Comment by mkosieradzki** [comment]
@yaron2 How about exposing opentracing-compatible API in the dapr sidecar? It would be great if it would support automated telemetry collectors like Application Insights - which are being ported to opentracing if I understand correctly.

**Comment by yaron2** [comment]
@mkosieradzki that's a great idea.

We have a plan to make telemetry collectors [Dapr components](https://github.com/dapr/components-contrib] for that purpose.

**Comment by LMWF** [comment]
Is this the same as this? 
https://github.com/dapr/components-contrib/tree/master/exporters

**Comment by mkosieradzki** [comment]
@LMWF Not really. What I mean is to expose unauthenticated endpoint in sidecar to which app can send it's internal telemetry, which dapr will forward using the exporter you mention.

**Comment by mmisztal1980** [comment]
I like the sidecar idea. Given the communication within a pod between it's containers transpires over the localhost interface, this is by far the quickest way to ingest telemetry via an unsecured endpoint.

**Comment by yaron2** [comment]
Yes. Now that we have separated exporters into the components-contrib repo, we are prime to start designing the app facing API.

**Comment by youngbupark** [comment]
Implementing collector is not easy, which can be conflicted with opentelemetry collector effort. Instead, we need to provide a way to correlate with user app tracing with dapr sidecar tracing. 

https://github.com/dapr/dapr/issues/1246

**Comment by karoldeland** [comment]
Would it be a good fit for logging application errors/exceptions?

**Comment by coip** [comment]
assuming this isn't stale, I think strides have been made with `otel` since.
might be worth incorporating here so you have the ability to plug into that much more.

eg, dapr-{`logs`,`traces`,`metrics`}.

otel docs for [logs](https://opentelemetry.io/docs/specs/otel/logs/)/[metrics](https://opentelemetry.io/docs/specs/otel/metrics/api/)/[traces](https://opentelemetry.io/docs/specs/otel/trace/api/)

Im assuming its crossed y'alls radar at some point, apologies if there's a more recent/true issue for this, just didn't see it. if that's the case tho, neat issue to close. 

spec -> https://opentelemetry.io/docs/specs/otel/
<details>

<summary>otel diagram</summary>

![](https://opentelemetry.io/img/architecture.png)

</details>
